components/script/dom/abstractworkerglobalscope.rs

```diff
@@ -13,12 +13,13 @@ use std::sync::mpsc::{Receiver, Sender};
 /// common event loop messages. While this SendableWorkerScriptChan is alive, the associated
 /// Worker object will remain alive.
 #[derive(Clone, JSTraceable)]
-pub struct SendableWorkerScriptChan<T: DomObject> {
+#[must_root]
+pub struct SendableWorkerScriptChan<#[must_root] T: DomObject> {
```

Comment:
I'm confused by these changes - `Trusted<T>` is a type that does not require rooting (due to [the allow_unrooted_interior attribute](https://github.com/servo/servo/blob/c9ba16f9fbdf7f43cb19feedfaaa68c85bbcbe3b/components/script/dom/bindings/refcounted.rs#L148)), so it's not clear why T needs to be marked must_root here.

Reply:
Sorry, I don't understand the relation between `Trusted<T>` and this place. I was putting `#[must_root]` in some places I thought it should be, without further understanding all of the code (there were too many errors).

For methods generic over T that we want to be able to accept `#[must_root]` we must specify explicitly that it is `#[must_root]` or mark the method with `allow_unrooted_interior` and/or `allow(unrooted_must_root)` (not sure which one).

Reply:
The reason I brought it up is that T is only used as an argument for the `Trusted<T>` member that is stored in SendableWorkerScriptChan. Since `Trusted<T>` is marked `#[allow_unrooted_interior]` it is allowed to be stored anywhere. SendableWorkerScriptChan should not be marked `#[must_root]` in that case. Is that case possible to represent with the current plugin?

Reply:
Oh, I see. I didn't notice the field 2 lines below. I got your point.

I don't remember how exactly all of these checks work. As I wrote somewhere else, here we are dealing with struct definition which are handled only on the HIR level if I remember correctly.

The intention with using `#[must_root]` is that it should be properly and explicitly propagated. So if we want `T` to be able to represent a `#[must_root]` type, then it should be marked as a `#[must_root]` type. Then, if we pass `T` to `Trusted<T>` and it is marked with `allow_unrooted_interior`, it means that `Trusted<T>` isn't `#[must_root]`, so the whole `SendableWorkerScriptChan` structure doesn't have to be `#[must_root]`, because it consists of regular fields.

Why is `SendableWorkerScriptChan` a `#[must_root]`? I marked it because there were some compilation errors, I believe. It is the heuristics I wrote about here and there.

There is also one side note: I am not sure if having `sth<T: U>` where `U` is `#[must_root]` enforces `T` to also be `#[must_root]`. But it should be possible to mute the error with `allow_unrooted_interior` and/or `allow(unrooted_must_root)`.